File: src/main/java/com/github/kindrat/cassandra/client/ui/window/editor/main/TableDataGridPane.java
package com.github.kindrat.cassandra.client.ui.window.editor.main;

import com.datastax.driver.core.TableMetadata;
import com.github.kindrat.cassandra.client.service.TableContext;
import com.github.kindrat.cassandra.client.ui.DataObject;
import com.github.kindrat.cassandra.client.ui.keylistener.TableCellCopyHandler;
import com.github.kindrat.cassandra.client.ui.window.editor.main.filter.FilterGrid;
import com.github.kindrat.cassandra.client.ui.window.editor.main.table.DataTableView;
import com.github.kindrat.cassandra.client.ui.window.editor.main.table.PaginationPanel;
import javafx.collections.ObservableList;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.input.KeyCode;
import javafx.scene.layout.ColumnConstraints;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import javafx.scene.layout.RowConstraints;
import reactor.core.publisher.EmitterProcessor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.FluxSink;

import java.util.List;

import static com.github.kindrat.cassandra.client.util.UIUtil.fillParent;
import static javafx.application.Platform.runLater;

public class TableDataGridPane extends GridPane {
    private final EmitterProcessor<DataObject> processor = EmitterProcessor.create();
    private final FluxSink<DataObject> sink = processor.sink();
    private final FilterGrid filterGrid;
    private final DataTableView dataTableView;
    private final PaginationPanel paginationPanel;

    public TableDataGridPane(FilterGrid filterGrid, DataTableView dataTableView, PaginationPanel paginationPanel) {
        this.filterGrid = filterGrid;
        this.dataTableView = dataTableView;
        this.paginationPanel = paginationPanel;
        setLayoutX(78);
        setLayoutY(131);
        setVisible(false);

        ColumnConstraints column = new ColumnConstraints();
        column.setHgrow(Priority.SOMETIMES);
        column.setMinWidth(10);
        column.setPrefWidth(100);

        getColumnConstraints().clear();
        getColumnConstraints().add(column);

        RowConstraints row1 = new RowConstraints();
        row1.setMaxHeight(40);
        row1.setMinHeight(40);
        row1.setPrefHeight(40);
        row1.setVgrow(Priority.SOMETIMES);

        RowConstraints row2 = new RowConstraints();
        row2.setVgrow(Priority.SOMETIMES);

        RowConstraints row3 = new RowConstraints();
        row3.setMaxHeight(40);
        row3.setMinHeight(40);
        row3.setPrefHeight(40);
        row3.setVgrow(Priority.SOMETIMES);

        getRowConstraints().clear();
        getRowConstraints().addAll(row1, row2, row3);

        add(filterGrid, 0, 0);
        add(dataTableView, 0, 1);
        add(paginationPanel, 0, 2);

        fillParent(this);
    }

    public void updateTableMetadata(TableMetadata metadata) {
        dataTableView.onTableSelected(metadata);
        filterGrid.setTableMetadata(metadata);
    }

    public void setDataColumns(List<TableColumn<DataObject, Object>> columns) {
        dataTableView.setDataColumns(columns);
    }

    public void setData(TableContext context, ObservableList<DataObject> data) {
        dataTableView.setItems(data);
        dataTableView.refresh();
        filterGrid.onDataUpdated(data);

        dataTableView.setOnKeyPressed(event -> {
            if (event.getCode() == KeyCode.DELETE) {
                TableView.TableViewSelectionModel<DataObject> selectionModel = dataTableView.getSelectionModel();
                DataObject selectedItem = selectionModel.getSelectedItem();
                int selectedIndex = selectionModel.getSelectedIndex();

                data.remove(selectedIndex);
                sink.next(selectedItem);
            }
        });

        paginationPanel.applyOnTable(context, dataObjects -> runLater(() -> {
            dataTableView.setItems(dataObjects);
            dataTableView.refresh();
        }));

        setVisible(true);
    }

    public Flux<DataObject> objectsToRemoveStream() {
        return processor;
    }

    public TableCellCopyHandler buildCopyHandler() {
        return dataTableView.buildCopyHandler();
    }

    public void suggestCompletion() {
        filterGrid.suggestCompletion();
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/window/editor/main/TableDataGridPane.kt
package com.github.kindrat.cassandra.client.ui.window.editor.main

import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata
import com.github.kindrat.cassandra.client.service.TableContext
import com.github.kindrat.cassandra.client.ui.DataObject
import com.github.kindrat.cassandra.client.ui.keylistener.TableCellCopyHandler
import com.github.kindrat.cassandra.client.ui.window.editor.main.filter.FilterGrid
import com.github.kindrat.cassandra.client.ui.window.editor.main.table.DataTableView
import com.github.kindrat.cassandra.client.ui.window.editor.main.table.PaginationPanel
import com.github.kindrat.cassandra.client.util.UIUtil.fillParent
import javafx.application.Platform
import javafx.collections.ObservableList
import javafx.event.EventHandler
import javafx.scene.control.TableColumn
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.ColumnConstraints
import javafx.scene.layout.GridPane
import javafx.scene.layout.Priority
import javafx.scene.layout.RowConstraints
import reactor.core.publisher.EmitterProcessor
import reactor.core.publisher.Flux
import java.util.function.Consumer

class TableDataGridPane(
        private val filterGrid: FilterGrid,
        private val dataTableView: DataTableView,
        private val paginationPanel: PaginationPanel
) : GridPane() {
    private val processor = EmitterProcessor.create<DataObject>()
    private val sink = processor.sink()

    fun updateTableMetadata(metadata: TableMetadata?) {
        dataTableView.onTableSelected(metadata!!)
        filterGrid.setTableMetadata(metadata)
    }

    fun setDataColumns(columns: List<TableColumn<DataObject, Any>>) {
        dataTableView.setDataColumns(columns)
    }

    fun setData(context: TableContext?, data: ObservableList<DataObject>) {
        dataTableView.items = data
        dataTableView.refresh()
        filterGrid.onDataUpdated(data)
        dataTableView.onKeyPressed = EventHandler { event: KeyEvent ->
            if (event.code == KeyCode.DELETE) {
                val selectionModel = dataTableView.selectionModel
                val selectedItem = selectionModel.selectedItem
                val selectedIndex = selectionModel.selectedIndex
                data.removeAt(selectedIndex)
                sink.next(selectedItem!!)
            }
        }
        paginationPanel.applyOnTable(context!!, Consumer {
            Platform.runLater {
                dataTableView.items = it
                dataTableView.refresh()
            }
        })
        isVisible = true
    }

    fun objectsToRemoveStream(): Flux<DataObject> {
        return processor
    }

    fun buildCopyHandler(): TableCellCopyHandler {
        return dataTableView.buildCopyHandler()
    }

    fun suggestCompletion() {
        filterGrid.suggestCompletion()
    }

    init {
        layoutX = 78.0
        layoutY = 131.0
        isVisible = false
        val column = ColumnConstraints()
        column.hgrow = Priority.SOMETIMES
        column.minWidth = 10.0
        column.prefWidth = 100.0
        columnConstraints.clear()
        columnConstraints.add(column)
        val row1 = RowConstraints()
        row1.maxHeight = 40.0
        row1.minHeight = 40.0
        row1.prefHeight = 40.0
        row1.vgrow = Priority.SOMETIMES
        val row2 = RowConstraints()
        row2.vgrow = Priority.SOMETIMES
        val row3 = RowConstraints()
        row3.maxHeight = 40.0
        row3.minHeight = 40.0
        row3.prefHeight = 40.0
        row3.vgrow = Priority.SOMETIMES
        rowConstraints.clear()
        rowConstraints.addAll(row1, row2, row3)
        add(filterGrid, 0, 0)
        add(dataTableView, 0, 1)
        add(paginationPanel, 0, 2)
        fillParent(this)
    }
}